### PR TITLE
HTTP POST producer error handling

### PR DESCRIPTION
**Pull request by stanlyDoge**
Bridge returns:
400 invalid request
422 POST request is empty
40401 topic does not exist
40402 partition does not exist (is higher than topic.partitions)

Not implemented "schema" error codes:
42201, 42202, 42205

### CODE REVIEW

File: src/main/java/io/strimzi/kafka/bridge/SourceBridgeEndpoint.java [hunk 1]
```diff
@@ -45,6 +50,10 @@
     private KafkaProducer<String, byte[]> producerUnsettledMode;
     private KafkaProducer<String, byte[]> producerSettledMode;
 
+    protected final String SEPARATOR = "#";
+    protected final int TOPIC_NOT_FOUND = 40401;
+    protected final int PARTITION_NOT_FOUND = 40402;
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/SourceBridgeEndpoint.java]: These are more specific to HTTP, can we move them to the related HTTP source bridge endpoint? Wdyt?

File: src/main/java/io/strimzi/kafka/bridge/SourceBridgeEndpoint.java [hunk 2]
```diff
@@ -80,11 +89,33 @@ protected void handleClose() {
...
 
-        log.debug("Sending to Kafka on topic {} at partition {} and key {}", krecord.topic(), krecord.partition(), krecord.key());
-        if (handler == null) {
-            this.producerSettledMode.write(krecord);
-        } else {
-            this.producerUnsettledMode.write(krecord, handler);
+        Properties props = new Properties();
+        KafkaConfig consumerConfig = this.bridgeConfigProperties.getKafkaConfig();
+        props.put("bootstrap.servers", consumerConfig.getBootstrapServers());
+        AdminClient ac = AdminClient.create(props);
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/SourceBridgeEndpoint.java]: not sure that we need the `AdminClient` which checks stuff before sending ... this will drive a slower client experience I guess. You should check what kind of exceptions are raised just sending a message when the above conditions happen (topic or partition don't exist). Finally, I am not sure it's the right place to do this check or at the upper level. Remember that this part is shared with the AMQP bridge so it should be agnostic. But we have to think more about that in terms of architecture.

File: src/main/java/io/strimzi/kafka/bridge/http/HttpBridge.java [hunk 3]
```diff
@@ -198,9 +198,18 @@ private void processRequests(HttpServerRequest httpServerRequest) {
                             .end();
                 }
                 break;
-
+            case EMPTY:
+                httpServerRequest.response()
+                        .setStatusCode(422)
+                        .setStatusMessage("records may not be empty")
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/http/HttpBridge.java]: "cannot be"

File: src/main/java/io/strimzi/kafka/bridge/http/HttpSourceBridgeEndpoint.java [hunk 4]
```diff
@@ -69,16 +69,18 @@ public void handle(Endpoint<?> endpoint) {
 
             // wait for ALL futures completed
             CompositeFuture.join(sendHandlers).setHandler(done -> {
+
                 for (int i = 0; i < sendHandlers.size(); i++) {
                     // check if, for each future, the sending operation is completed successfully or failed
-                    if (done.result().succeeded(i)) {
+                    if (done.result() != null && done.result().succeeded(i)) {
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/http/HttpSourceBridgeEndpoint.java]: good catch!

File: src/test/java/io/strimzi/kafka/bridge/amqp/AmqpBridgeTest.java [hunk 5]
```diff
@@ -109,6 +109,13 @@ public void sendSimpleMessages(TestContext context) {
 		String topic = "sendSimpleMessages";
 		kafkaCluster.createTopic(topic, 1, 1);
 
+		try {
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/amqp/AmqpBridgeTest.java]: why the need for sleep now? The tests were all passing up to now.

File: src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeErrorsTest.java [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+        this.vertx.deployVerticle(this.httpBridge, context.asyncAssertSuccess());
+    }
+
+    @After
+    public void after(TestContext context) {
+
+        this.vertx.close(context.asyncAssertSuccess());
+    }
+
+    //@Test
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeErrorsTest.java]: is it disabled on purpose?

File: src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeTest.java [hunk 7]
```diff
@@ -74,6 +74,12 @@ public void before(TestContext context) {
         this.httpBridge = new HttpBridge(this.bridgeConfigProperties);
 
         this.vertx.deployVerticle(this.httpBridge, context.asyncAssertSuccess());
+        try {
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeTest.java]: do you really need this sleep? Tests are passing even on Travis without that, are you facing any problems?

File: src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeTest.java [hunk 8]
```diff
@@ -742,4 +773,44 @@ public void receiveSimpleMessageFromPartitionAndOffset(TestContext context) {
 
         deleteAsync.await();
     }
+
+    //@Test
+    public void sendToOnePartitionTest(TestContext context) {
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeTest.java]: is it disabled on purpose?

File: src/test/java/io/strimzi/kafka/bridge/amqp/AmqpBridgeTest.java (hunk 5 above)
Comment (stanlyDoge) [reply src/test/java/io/strimzi/kafka/bridge/amqp/AmqpBridgeTest.java]: I turned off the autocreation of the topics to test non-existing topics. Before, it just created a topic when a message was sent to it.

File: src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeTest.java (hunk 8 above)
Comment (stanlyDoge) [reply src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeTest.java]: Yeah. From the API spec I see `POST /topics/(string: topic_name)/partitions/(int: partition_id)` request should do `"Produce messages to one partition of the topic"` but in our bridge, it is fully ignored (AFAICS). So this is the test which we can allow after this is implemented.

Comment (ppatierno) [reply src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeTest.java]: But using that endpoint is another way to do that. You can already send to one partition when you specify "partition" field in a record in the "records" collection when you do a POST on `/topics/topic_name`

Comment (ppatierno) [reply src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeTest.java]: Ah! I think I got it. You added test but not implementing the feature yet I guess. Well, I would avoid that, it creates just confusion imho. This PR is about producer error handling not implementing that endpoint.

Comment (stanlyDoge) [reply src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeTest.java]: Ok, I just wanted to write a test which uses this endpoint. And I did O:) However, it does not cope properly with the situation which is described as "If no partition is provided, one will be chosen based on the hash of the key. If no key is provided, the partition will be chosen for each message in a round-robin fashion."

Comment (ppatierno) [reply src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeTest.java]: That description works with the generic `/topics/topic-name` endpoint of course, not this one.

File: src/main/java/io/strimzi/kafka/bridge/SourceBridgeEndpoint.java [hunk 9]
```diff
@@ -80,11 +82,25 @@ protected void handleClose() {
...
         if (handler == null) {
             this.producerSettledMode.write(krecord);
         } else {
-            this.producerUnsettledMode.write(krecord, handler);
+            log.debug("Sending to topic " + krecord.topic() + " at partition {}", krecord.partition());
+            this.producerUnsettledMode.partitionsFor(krecord.topic(), part -> {
+                if (part.succeeded()) {
+                    if (krecord.partition() == null) {
+                        //TODO should not to be!
+                        this.producerUnsettledMode.write(krecord, handler);
```
Comment (stanlyDoge) [inline src/main/java/io/strimzi/kafka/bridge/SourceBridgeEndpoint.java]: So you say this is applying round-robin somehow and I can just delete this comment? @ppatierno

Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/SourceBridgeEndpoint.java]: When the partition is null and the key is null, the producer uses round robin but it's not a logic we have to add it's in the Kafka producer. The code here hasn't to be changed. You should try to catch the right exception for these kinds of errors. Please try and see the right exception raised by the producer.

Comment (stanlyDoge) [reply src/main/java/io/strimzi/kafka/bridge/SourceBridgeEndpoint.java]: In that case, everything is ok :)

File: src/main/java/io/strimzi/kafka/bridge/SourceBridgeEndpoint.java [hunk 10]
```diff
@@ -86,21 +86,7 @@ protected void send(KafkaProducerRecord<String, byte[]> krecord, Handler<AsyncRe
...
-                        this.producerUnsettledMode.write(krecord, handler);
-
-                    } else {
-                        handler.handle(Future.failedFuture(Integer.toString(ErrorCodeEnum.PARTITION_NOT_FOUND.getValue())));
-                    }
-                } else {
-                    handler.handle(Future.failedFuture(Integer.toString(ErrorCodeEnum.TOPIC_NOT_FOUND.getValue())));
-                }
-            });
+            this.producerUnsettledMode.write(krecord, handler).exceptionHandler(k -> handler.handle(Future.failedFuture(k)));
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/SourceBridgeEndpoint.java]: I would change the order setting the exception handler first and before actually writing.

File: src/main/java/io/strimzi/kafka/bridge/SourceBridgeEndpoint.java [hunk 11]
```diff
@@ -80,11 +82,11 @@ protected void handleClose() {
      */
     protected void send(KafkaProducerRecord<String, byte[]> krecord, Handler<AsyncResult<RecordMetadata>> handler) {
 
-        log.debug("Sending to Kafka on topic {} at partition {} and key {}", krecord.topic(), krecord.partition(), krecord.key());
         if (handler == null) {
             this.producerSettledMode.write(krecord);
         } else {
-            this.producerUnsettledMode.write(krecord, handler);
+            log.debug("Sending to topic " + krecord.topic() + " at partition {}", krecord.partition());
+            this.producerUnsettledMode.exceptionHandler(e -> handler.handle(Future.failedFuture(e))).write(krecord, handler);
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/SourceBridgeEndpoint.java]: I have just a doubt but you could check easily. Are we sure that the `handler` isn't called twice?
So one time by the exceptionHandler and the other time by the write but with the async result as failed (which is something we check in the handler).

File: src/main/java/io/strimzi/kafka/bridge/SourceBridgeEndpoint.java [hunk 12]
```diff
@@ -96,6 +98,7 @@ public void open() {
         props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, this.bridgeConfigProperties.getKafkaConfig().getProducerConfig().getKeySerializer());
         props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, this.bridgeConfigProperties.getKafkaConfig().getProducerConfig().getValueSerializer());
         props.put(ProducerConfig.ACKS_CONFIG, this.bridgeConfigProperties.getKafkaConfig().getProducerConfig().getAcks());
+        props.put(ProducerConfig.MAX_BLOCK_MS_CONFIG, 10000);
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/SourceBridgeEndpoint.java]: why this one?

File: src/main/java/io/strimzi/kafka/bridge/http/HttpBridge.java [hunk 13]
```diff
@@ -194,13 +194,23 @@ private void processRequests(HttpServerRequest httpServerRequest) {
...
+                            .setStatusCode(ErrorCodeEnum.UNKNOWN_REQUEST.getValue())
+                            .setStatusMessage("endpoint not found")
                             .end();
                 }
                 break;
-
+            case EMPTY:
+                httpServerRequest.response()
+                        .setStatusCode(ErrorCodeEnum.EMPTY_REQUEST.getValue())
+                        .setStatusMessage("records cannot not be empty")
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/http/HttpBridge.java]: we don't know which kind of request this is. Cannot say nothing about the payload and what it should contain. It could be a POST request related to commit offsets so it doesn't contain records.

File: src/main/java/io/strimzi/kafka/bridge/http/ErrorCodeEnum.java [hunk 14]
```diff
@@ -0,0 +1,34 @@
+package io.strimzi.kafka.bridge.http;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public enum ErrorCodeEnum {
+    TOPIC_NOT_FOUND(40401),
+    PARTITION_NOT_FOUND(40402),
+    INVALID_REQUEST(400),
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/http/ErrorCodeEnum.java]: let's use HTTP name so "bad request"

File: src/main/java/io/strimzi/kafka/bridge/http/ErrorCodeEnum.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    UNKNOWN_REQUEST(404),
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/http/ErrorCodeEnum.java]: not found

File: src/main/java/io/strimzi/kafka/bridge/http/ErrorCodeEnum.java [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+    EMPTY_REQUEST(422),
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/http/ErrorCodeEnum.java]: unprocessable entity

File: src/main/java/io/strimzi/kafka/bridge/http/ErrorCodeEnum.java [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+    UNKNOWN_ERROR(500);
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/http/ErrorCodeEnum.java]: internal server error

File: src/main/java/io/strimzi/kafka/bridge/http/HttpSourceBridgeEndpoint.java [hunk 18]
```diff
@@ -47,16 +47,21 @@ public HttpSourceBridgeEndpoint(Vertx vertx, HttpBridgeConfig bridgeConfigProper
...
-        messageConverter = new HttpJsonMessageConverter();
 
         // split path to extract params
         String[] params = httpServerRequest.path().split("/");
 
         // path is like this : /topics/{topic_name}, topic will be at the last position of params
-        String topic = params[params.length - 1];
+        String topic = params[2];
+
+        if (params.length == 5) {
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/http/HttpSourceBridgeEndpoint.java]: Why this distinction? what is param number 4? If it's about partition-id on the URL we should remove that, it's not part of this PR.

File: src/main/java/io/strimzi/kafka/bridge/http/converter/HttpJsonMessageConverter.java [hunk 19]
```diff
@@ -30,6 +30,17 @@
...
+    private Integer partition;
+
+    public HttpJsonMessageConverter(int partition) {
+        this.partition = partition;
+    }
+
+    public HttpJsonMessageConverter() {
+        super();
+        this.partition = null;
+    }
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/http/converter/HttpJsonMessageConverter.java]: as said partition-id on the URL is not part of this PR, please remove all the stuff related to it.

File: src/main/java/io/strimzi/kafka/bridge/http/converter/HttpJsonMessageConverter.java [hunk 20]
```diff
@@ -38,14 +49,29 @@
...
+                // TODO
+                // If no partition is provided, one will be chosen based on the hash of the key.
+                // If no key is provided, the partition will be chosen for each message in a round-robin fashion.
+                /*if (key != null) {
+                    partition = key.hashCode();
+                } else {
+                    partition = roundRobin();
+                }*/
+                partition = 0;
+            }
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/http/converter/HttpJsonMessageConverter.java]: is that the same here? logic for partition-id on the URL?

File: src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeErrorsTest.java [hunk 21]
```diff
@@ -0,0 +1,266 @@
...
+                    String statusMessage = offsets.getJsonObject(0).getString("error");
+
+                    context.assertEquals(ErrorCodeEnum.TOPIC_NOT_FOUND.getValue(), code);
+                    context.assertEquals("Topic " + kafkaTopic + " not found", statusMessage);
+                    async.complete();
+                });
+    }
+
+    //@Test
+    public void sendToNonExistingTopicNonExistingPartitionTest(TestContext context) {
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeErrorsTest.java]: instead of commenting, let's use @Ignore so that we have trace during the tests that something is ignored and it the future we can re-activate that.

File: src/main/java/io/strimzi/kafka/bridge/SourceBridgeEndpoint.java [hunk 22]
```diff
@@ -80,11 +81,13 @@ protected void handleClose() {
      */
     protected void send(KafkaProducerRecord<String, byte[]> krecord, Handler<AsyncResult<RecordMetadata>> handler) {
 
-        log.debug("Sending to Kafka on topic {} at partition {} and key {}", krecord.topic(), krecord.partition(), krecord.key());
         if (handler == null) {
             this.producerSettledMode.write(krecord);
         } else {
-            this.producerUnsettledMode.write(krecord, handler);
+            log.debug("Sending to topic " + krecord.topic() + " at partition {}", krecord.partition());
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/SourceBridgeEndpoint.java]: let's move this log line before the `if` statement, because it's the same for both producers.

File: src/main/java/io/strimzi/kafka/bridge/http/HttpBridge.java [hunk 23]
```diff
@@ -194,13 +194,23 @@ private void processRequests(HttpServerRequest httpServerRequest) {
...
+                            .setStatusCode(ErrorCodeEnum.NOT_FOUND.getValue())
+                            .setStatusMessage("endpoint not found")
                             .end();
                 }
                 break;
-
+            case EMPTY:
+                httpServerRequest.response()
+                        .setStatusCode(ErrorCodeEnum.UNPROCESSABLE_ENTITY.getValue())
+                        .setStatusMessage("records cannot not be empty")
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/http/HttpBridge.java]: I still see "records" in this message. But it could be not a POST request about records but about offsets for example.

File: src/main/java/io/strimzi/kafka/bridge/http/HttpSourceBridgeEndpoint.java [hunk 24]
```diff
@@ -47,16 +47,17 @@ public HttpSourceBridgeEndpoint(Vertx vertx, HttpBridgeConfig bridgeConfigProper
...
         HttpServerRequest httpServerRequest = (HttpServerRequest) endpoint.get();
 
-        messageConverter = new HttpJsonMessageConverter();
 
         // split path to extract params
         String[] params = httpServerRequest.path().split("/");
 
         // path is like this : /topics/{topic_name}, topic will be at the last position of params
-        String topic = params[params.length - 1];
+        String topic = params[2];
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/http/HttpSourceBridgeEndpoint.java]: you are still thinking about having more after the topic_name (so the partition-id) ... please revert this as well :-)

File: src/main/java/io/strimzi/kafka/bridge/http/RequestIdentifier.java [hunk 25]
```diff
@@ -44,6 +45,9 @@ static RequestType getRequestType(HttpServerRequest request){
                 if (method == HttpMethod.POST && params.length == 2){
                     return RequestType.PRODUCE;
                 }
+                if (method == HttpMethod.POST && params.length == 4 && params[2].equals("partitions") && isNumeric(params[3])){
+                    return RequestType.PRODUCE;
+                }
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/http/RequestIdentifier.java]: still a piece of partition-id in the URL :-)

File: src/main/java/io/strimzi/kafka/bridge/http/converter/HttpJsonMessageConverter.java [hunk 26]
```diff
@@ -30,6 +30,10 @@
 
 public class HttpJsonMessageConverter implements MessageConverter<String, byte[], Buffer, Buffer> {
 
+    public HttpJsonMessageConverter() {
+        super();
+    }
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/http/converter/HttpJsonMessageConverter.java]: why do we need this ctor?

File: src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeErrorsTest.java [hunk 27]
```diff
@@ -0,0 +1,270 @@
...
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+
+@RunWith(VertxUnitRunner.class)
+public class HttpBridgeErrorsTest extends KafkaClusterTestBase {
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeErrorsTest.java]: do we really need a complete new class for testing errors. Imho we can just having tests about errors in the current class we already have.

File: src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeTest.java [hunk 28]
```diff
@@ -742,4 +748,285 @@ public void receiveSimpleMessageFromPartitionAndOffset(TestContext context) {
...
+                    context.assertEquals(0L, metadata.getLong("offset"));
+                    async.complete();
+                });
+    }
+
+    @Ignore
+    @Test
+    public void partitionPriorityTest(TestContext context) {
+        // partition is specified in the message body and in the POST request
+        // what does have higher priority? Lets say POST one
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeTest.java]: you are right that the priority is on the URL because the client is doing a POST to a specific endpoint but I would return an error about wrong payload because you cannot specify the "partition" in the JSON. Pushing silently, can confuse the user which thinks that we pushed the message to the partition in the JSON while it went to the one in the URL.
